### PR TITLE
metis: reduce MAX_JBUFS to keep clang 7.3 happy

### LINKED ISSUE #3461: Error installing metis
Hi all,

I have an issue installing `metis` on OSX 10.11.2.

I am using the following `clang`:

```bash
$ clang -v
Apple LLVM version 7.3.0 (clang-703.0.29)
Target: x86_64-apple-darwin15.2.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

Steps to reproduce:

```bash
$ brew tap homebrew/science
$ brew install metis
==> Installing metis from homebrew/science
==> Downloading http://glaros.dtc.umn.edu/gkhome/fetch/sw/metis/metis-5.1.0.tar.gz
Already downloaded: /Library/Caches/Homebrew/metis-5.1.0.tar.gz
==> make config shared=1 prefix=/usr/local/Cellar/metis/5.1.0 openmp=0
==> make install
Last 15 lines from /Users/irisvanzelst/Library/Logs/Homebrew/metis/02.make:
[ 67%] Building C object libmetis/CMakeFiles/metis.dir/timing.c.o
cd /tmp/metis20160322-8216-15gx1te/metis-5.1.0/build/Darwin-x86_64/libmetis && /usr/local/Library/ENV/4.3/clang  -Dmetis_EXPORTS -I/tmp/metis20160322-8216-15gx1te/metis-5.1.0/GKlib -I/tmp/metis20160322-8216-15gx1te/metis-5.1.0/include -I/tmp/metis20160322-8216-15gx1te/metis-5.1.0/libmetis/.  -DLINUX -D_FILE_OFFSET_BITS=64 -DNDEBUG -DNDEBUG2 -DHAVE_EXECINFO_H -DHAVE_GETLINE -O3 -fPIC   -o CMakeFiles/metis.dir/timing.c.o   -c /tmp/metis20160322-8216-15gx1te/metis-5.1.0/libmetis/timing.c
[ 68%] Building C object libmetis/CMakeFiles/metis.dir/util.c.o
cd /tmp/metis20160322-8216-15gx1te/metis-5.1.0/build/Darwin-x86_64/libmetis && /usr/local/Library/ENV/4.3/clang  -Dmetis_EXPORTS -I/tmp/metis20160322-8216-15gx1te/metis-5.1.0/GKlib -I/tmp/metis20160322-8216-15gx1te/metis-5.1.0/include -I/tmp/metis20160322-8216-15gx1te/metis-5.1.0/libmetis/.  -DLINUX -D_FILE_OFFSET_BITS=64 -DNDEBUG -DNDEBUG2 -DHAVE_EXECINFO_H -DHAVE_GETLINE -O3 -fPIC   -o CMakeFiles/metis.dir/util.c.o   -c /tmp/metis20160322-8216-15gx1te/metis-5.1.0/libmetis/util.c
[ 69%] Building C object libmetis/CMakeFiles/metis.dir/wspace.c.o
cd /tmp/metis20160322-8216-15gx1te/metis-5.1.0/build/Darwin-x86_64/libmetis && /usr/local/Library/ENV/4.3/clang  -Dmetis_EXPORTS -I/tmp/metis20160322-8216-15gx1te/metis-5.1.0/GKlib -I/tmp/metis20160322-8216-15gx1te/metis-5.1.0/include -I/tmp/metis20160322-8216-15gx1te/metis-5.1.0/libmetis/.  -DLINUX -D_FILE_OFFSET_BITS=64 -DNDEBUG -DNDEBUG2 -DHAVE_EXECINFO_H -DHAVE_GETLINE -O3 -fPIC   -o CMakeFiles/metis.dir/wspace.c.o   -c /tmp/metis20160322-8216-15gx1te/metis-5.1.0/libmetis/wspace.c
[ 70%] Linking C shared library libmetis.dylib
cd /tmp/metis20160322-8216-15gx1te/metis-5.1.0/build/Darwin-x86_64/libmetis && /usr/local/Cellar/cmake/3.5.0/bin/cmake -E cmake_link_script CMakeFiles/metis.dir/link.txt --verbose=1
/usr/local/Library/ENV/4.3/clang   -DLINUX -D_FILE_OFFSET_BITS=64 -DNDEBUG -DNDEBUG2 -DHAVE_EXECINFO_H -DHAVE_GETLINE -O3 -dynamiclib -Wl,-headerpad_max_install_names  -o libmetis.dylib -install_name /tmp/metis20160322-8216-15gx1te/metis-5.1.0/build/Darwin-x86_64/libmetis/libmetis.dylib CMakeFiles/metis.dir/__/GKlib/b64.c.o CMakeFiles/metis.dir/__/GKlib/blas.c.o CMakeFiles/metis.dir/__/GKlib/csr.c.o CMakeFiles/metis.dir/__/GKlib/error.c.o CMakeFiles/metis.dir/__/GKlib/evaluate.c.o CMakeFiles/metis.dir/__/GKlib/fkvkselect.c.o CMakeFiles/metis.dir/__/GKlib/fs.c.o CMakeFiles/metis.dir/__/GKlib/getopt.c.o CMakeFiles/metis.dir/__/GKlib/gkregex.c.o CMakeFiles/metis.dir/__/GKlib/graph.c.o CMakeFiles/metis.dir/__/GKlib/htable.c.o CMakeFiles/metis.dir/__/GKlib/io.c.o CMakeFiles/metis.dir/__/GKlib/itemsets.c.o CMakeFiles/metis.dir/__/GKlib/mcore.c.o CMakeFiles/metis.dir/__/GKlib/memory.c.o CMakeFiles/metis.dir/__/GKlib/omp.c.o CMakeFiles/metis.dir/__/GKlib/pdb.c.o CMakeFiles/metis.dir/__/GKlib/pqueue.c.o CMakeFiles/metis.dir/__/GKlib/random.c.o CMakeFiles/metis.dir/__/GKlib/rw.c.o CMakeFiles/metis.dir/__/GKlib/seq.c.o CMakeFiles/metis.dir/__/GKlib/sort.c.o CMakeFiles/metis.dir/__/GKlib/string.c.o CMakeFiles/metis.dir/__/GKlib/timers.c.o CMakeFiles/metis.dir/__/GKlib/tokenizer.c.o CMakeFiles/metis.dir/__/GKlib/util.c.o CMakeFiles/metis.dir/auxapi.c.o CMakeFiles/metis.dir/balance.c.o CMakeFiles/metis.dir/bucketsort.c.o CMakeFiles/metis.dir/checkgraph.c.o CMakeFiles/metis.dir/coarsen.c.o CMakeFiles/metis.dir/compress.c.o CMakeFiles/metis.dir/contig.c.o CMakeFiles/metis.dir/debug.c.o CMakeFiles/metis.dir/fm.c.o CMakeFiles/metis.dir/fortran.c.o CMakeFiles/metis.dir/frename.c.o CMakeFiles/metis.dir/gklib.c.o CMakeFiles/metis.dir/graph.c.o CMakeFiles/metis.dir/initpart.c.o CMakeFiles/metis.dir/kmetis.c.o CMakeFiles/metis.dir/kwayfm.c.o CMakeFiles/metis.dir/kwayrefine.c.o CMakeFiles/metis.dir/mcutil.c.o CMakeFiles/metis.dir/mesh.c.o CMakeFiles/metis.dir/meshpart.c.o CMakeFiles/metis.dir/minconn.c.o CMakeFiles/metis.dir/mincover.c.o CMakeFiles/metis.dir/mmd.c.o CMakeFiles/metis.dir/ometis.c.o CMakeFiles/metis.dir/options.c.o CMakeFiles/metis.dir/parmetis.c.o CMakeFiles/metis.dir/pmetis.c.o CMakeFiles/metis.dir/refine.c.o CMakeFiles/metis.dir/separator.c.o CMakeFiles/metis.dir/sfm.c.o CMakeFiles/metis.dir/srefine.c.o CMakeFiles/metis.dir/stat.c.o CMakeFiles/metis.dir/timing.c.o CMakeFiles/metis.dir/util.c.o CMakeFiles/metis.dir/wspace.c.o -lm 
ld: section __DATA/__thread_bss extends beyond end of file, file 'CMakeFiles/metis.dir/__/GKlib/error.c.o' for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [libmetis/libmetis.dylib] Error 1
make[2]: *** [libmetis/CMakeFiles/metis.dir/all] Error 2
make[1]: *** [all] Error 2
make: *** [install] Error 2
```

Output of `brew doctor` (should hopefully all be unrelated):

```bash
 brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry and just ignore them. Thanks!

Warning: Some directories in /usr/local/share/man aren't writable.
This can happen if you "sudo make install" software that isn't managed
by Homebrew. If a brew tries to add locale information to one of these
directories, then the install will fail during the link step.

You should probably `sudo chown -R $(whoami)` them:
    /usr/local/share/man/de
    /usr/local/share/man/de/man1
    /usr/local/share/man/mann

Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected dylibs:
    /usr/local/lib/libtcl8.6.dylib
    /usr/local/lib/libtk8.6.dylib

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
    /usr/local/include/fakemysql.h
    /usr/local/include/fakepq.h
    /usr/local/include/fakesql.h
    /usr/local/include/itcl.h
    /usr/local/include/itcl2TclOO.h
    /usr/local/include/itclDecls.h
    /usr/local/include/itclInt.h
    /usr/local/include/itclIntDecls.h
    /usr/local/include/itclMigrate2TclCore.h
    /usr/local/include/itclTclIntStubsFcn.h
    /usr/local/include/mysqlStubs.h
    /usr/local/include/odbcStubs.h
    /usr/local/include/pqStubs.h
    /usr/local/include/tcl.h
    /usr/local/include/tclDecls.h
    /usr/local/include/tclOO.h
    /usr/local/include/tclOODecls.h
    /usr/local/include/tclPlatDecls.h
    /usr/local/include/tclThread.h
    /usr/local/include/tclTomMath.h
    /usr/local/include/tclTomMathDecls.h
    /usr/local/include/tdbc.h
    /usr/local/include/tdbcDecls.h
    /usr/local/include/tdbcInt.h
    /usr/local/include/tk.h
    /usr/local/include/tkDecls.h
    /usr/local/include/tkPlatDecls.h

Warning: Unbrewed .pc files were found in /usr/local/lib/pkgconfig.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected .pc files:
    /usr/local/lib/pkgconfig/tcl.pc
    /usr/local/lib/pkgconfig/tk.pc

Warning: Unbrewed static libraries were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected static libraries:
    /usr/local/lib/libtclstub8.6.a
    /usr/local/lib/libtkstub8.6.a

Warning: Broken symlinks were found. Remove them with `brew prune`:
    /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraKakuPro-W3
    /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraKakuPro-W6
    /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraKakuProN-W3
    /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraKakuProN-W6
    /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraKakuStd-W8
    /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraKakuStdN-W8
    /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraMaruPro-W4
    /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraMaruProN-W4
    /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraMinPro-W3
    /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraMinPro-W6
    /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraMinProN-W3
    /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraMinProN-W6
    /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraginoSansGB-W3
    /usr/local/share/ghostscript/9.16/Resource/CIDFont/HiraginoSansGB-W6
    /usr/local/share/ghostscript/9.16/Resource/CIDFont/YuMin-Demibold
    /usr/local/share/ghostscript/9.16/Resource/CIDFont/YuMin-Medium
```

Any ideas what the issue here is? Thanks for your help!

### PR DESCRIPTION
**Pull request by ilovezfs**
- clang 7.3 requires MAX_JBUFS < 25 for a successful build.
- The precise tipping point was determined by trial and error.

Closes #3461

### CODE REVIEW

File: metis.rb [hunk 1]
```diff
@@ -16,6 +17,12 @@ class Metis < Formula
 
   def install
     ENV.universal_binary if build.universal?
+
+    # Make clang 7.3 happy. Check if still needed when Xcode > 7.3 is released.
+    # Prevents "ld: section __DATA/__thread_bss extends beyond end of file"
+    # See upstream LLVM issue https://llvm.org/bugs/show_bug.cgi?id=27059
+    inreplace "GKlib/error.c", "#define MAX_JBUFS 128", "#define MAX_JBUFS 24"
```
Comment (xu-cheng) [inline metis.rb]: Shouldn't this be submit to Metis's upstream as well?

**Comment by xu-cheng** [comment]
I'm not too familiar with this. But :+1: if it works.

File: metis.rb (hunk 1 above)
Comment (ilovezfs) [inline metis.rb]: It's a clang bug that LLVM should fix.

Comment (ilovezfs) [inline metis.rb]: I'll send an email to Professor Karypis letting him know as well.

**Comment by UniqMartin** [comment]
Unless you know exactly what `MAX_JBUFS` is doing, I think it's a very bad idea to just change that from what the original authors of the package set it to. (The comment notes it has been reported, but has it also been confirmed by upstream?)

*Returning back to the actual issue:*

```
ld: section __DATA/__thread_bss extends beyond end of file
```

This sounds more like a linker bug than a compiler bug, because there's nothing wrong with a BSS section extending beyond the end of the file. This is a section with all-zero data and there's no need to waste disk space for it. Like other BSS-type sections it is not expected to be backed by physical storage and thus its bounds (almost) don't matter. Would be interesting to see what changed from Xcode 7.2 to 7.3, or what these and GCC produce for the problematic file (sections can be inspected with `otool -l`).

**Comment by ilovezfs** [comment]
@UniqMartin 
```
/* These are the jmp_buf for the graceful exit in case of severe errors.
   Multiple buffers are defined to allow for recursive invokation. */
#define MAX_JBUFS 128
__thread int gk_cur_jbufs=-1;
__thread jmp_buf gk_jbufs[MAX_JBUFS];
__thread jmp_buf gk_jbuf;

```
Not a problem.

**Comment by UniqMartin** [comment]
Good; it seems to be safe in this case. Just wanted to leave a word of caution just in case.

**Comment by ilovezfs** [comment]
@UniqMartin Yup, understood. I was wary, too, till I looked under the hood.

**Pull request closed without merging**